crates/biome_js_analyze/src/lint/nursery/use_readonly_class_properties.rs

```diff
@@ -519,122 +562,88 @@ fn collect_mutated_class_property_names(members: &JsClassMemberList) -> Vec<Text
 /// Checks recursively the assignment operand equals a reference to `this` (e.g. `this.privateProp`)
 fn contains_this_or_static_member_kind(
     assignment: &JsStaticMemberAssignment,
-    this_aliases: &[Text],
+    this_aliases: &[ThisAliasesAndTheirScope],
 ) -> bool {
     if let Ok(object) = assignment.object() {
-        let js_this_assignment = object.as_js_this_expression();
-
-        if js_this_assignment.is_some() {
+        if object.as_js_this_expression().is_some() {
             return true;
         }
 
         if let Some(js_identifier_expression) = object.as_js_identifier_expression() {
             if let Ok(name) = js_identifier_expression.name() {
+                let name_text = name.to_trimmed_text();
+                let name_syntax = name.syntax();
+
                 return this_aliases
                     .iter()
-                    .any(|alias| alias.eq(&name.to_trimmed_text()));
+                    .any(|ThisAliasesAndTheirScope { aliases, scope }| {
+                        aliases.contains(&name_text)
+                            && name_syntax
+                                .ancestors()
+                                .any(|ancestor| ancestor.key() == scope.syntax().key())
+                    });
             }
         }
     }
 
     false
 }
 
-/// Recursively traverses a class method or setter body (or related AST node)
-/// and collects the names of all properties assigned to this (or its aliases) within that body.
-/// It handles assignment expressions, update expressions, array/object destructuring, and nested method bodies,
-/// returning an iterator of property names (Text) that are mutated.
-/// This is used to detect which class properties are reassigned inside methods or setters.
-fn collect_all_assignment_names(
+fn visit_fn_body_descendants<F>(
     method_body_element: &MethodBodyElementOrStatementList,
-    this_aliases: &[Text],
-) -> impl Iterator<Item = Text> {
-    method_body_element
-        .syntax()
-        .children()
-        .flat_map(move |child| {
-            // Try to handle assignment expressions
-            if let Some(expr) = JsAssignmentExpression::cast_ref(&child) {
-                if let Ok(left) = expr.left() {
-                    // Handle array assignment pattern
-                    if let Some(assignment) = left.as_js_array_assignment_pattern().cloned() {
-                        return extract_js_array_assignment_pattern_names(
-                            &assignment,
-                            this_aliases,
-                        )
-                        .into_iter();
+    this_aliases: &[ThisAliasesAndTheirScope],
+    on_name: &mut F,
+) where
+    F: FnMut(Text),
+{
+    method_body_element.syntax().children().for_each(|child| {
+        if let Some(expr) = JsAssignmentExpression::cast_ref(&child) {
+            if let Ok(left) = expr.left() {
+                if let Some(assignment) = left.as_js_array_assignment_pattern().cloned() {
```

Comment:
Can we use let chains to avoid this nesting? https://rust-for-c-programmers.com/ch21/21_15_if_let_chains_rust_2024.html